### PR TITLE
Remove net451 as a cross-compile target

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: test/Microsoft.AspNetCore.Mvc.Core.Test/NonDisposableStreamTest.cs [hunk 1]
```diff
@@ -23,6 +23,21 @@ public void InnerStreamIsOpenOnClose()
...
+            var stream = FlushReportingStream.GetThrowingStream();
+
+            var nonDisposableStream = new NonDisposableStream(stream);
+
+            // Act & Assert
+            nonDisposableStream.Close();
+        }
+#elif NETCOREAPP2_0
+#else
+    #error The target frameworks need to be updated
```
Comment (kichalla) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/NonDisposableStreamTest.cs]: nit: remove indentation

File: samples/MvcSandbox/MvcSandbox.csproj [hunk 2]
```diff
@@ -3,8 +3,8 @@
   <Import Project="..\..\build\common.props" />
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp1.1;net452</TargetFrameworks>
-    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netcoreapp1.1</TargetFrameworks>
+    <TargetFrameworks>netcoreapp2.0;net46</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netcoreapp2.0</TargetFrameworks>
```
Comment (dougbu) [inline samples/MvcSandbox/MvcSandbox.csproj]: Suggest adding a `$(NetCoreAppVersion)` property in dependency.props.

File: src/Microsoft.AspNetCore.Mvc.ApiExplorer/Microsoft.AspNetCore.Mvc.ApiExplorer.csproj [hunk 3]
```diff
@@ -4,7 +4,7 @@
 
   <PropertyGroup>
     <Description>ASP.NET Core MVC API explorer functionality for discovering metadata such as the list of controllers and actions, and their URLs and allowed HTTP methods.</Description>
-    <TargetFrameworks>net451;netstandard1.6</TargetFrameworks>
+    <TargetFrameworks>net46;netstandard1.6</TargetFrameworks>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ApiExplorer/Microsoft.AspNetCore.Mvc.ApiExplorer.csproj]: Why do some product packages still have multiple targets?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/EmptyArray.cs [hunk 4]
```diff
@@ -1,19 +1,12 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-#if !NET451
 using System;
-#endif
 
 namespace Microsoft.AspNetCore.Mvc.Internal
 {
     public static class EmptyArray<TElement>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/EmptyArray.cs]: Should just remove this class and update places it's used. Class existed only to work around the .NET 4.5's lack of `Array.Empty<T>()`.

File: src/Microsoft.AspNetCore.Mvc.Formatters.Xml/Internal/FormattingUtilities.cs [hunk 5]
```diff
@@ -15,8 +15,11 @@ public static class FormattingUtilities
     {
         public static readonly int DefaultMaxDepth = 32;
 
-#if NET451
+#if NET46
         public static readonly XsdDataContractExporter XsdDataContractExporter = new XsdDataContractExporter();
+#elif NETSTANDARD1_6
+#else
+#error the target framework needs to be updated.
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Xml/Internal/FormattingUtilities.cs]: Right now, the `#if`'s look pretty random. Do this `#error` either once per project or in every C# file containing `#if`s.

File: src/Microsoft.AspNetCore.Mvc.WebApiCompatShim/ContentNegotiator/FormattingUtilities.cs [hunk 6]
```diff
@@ -162,9 +162,6 @@ public static XmlDictionaryReaderQuotas CreateDefaultReaderQuotas()
         {
             // MaxDepth is a DOS mitigation. We don't support MaxDepth in portable libraries because it is strictly
             // client side.
-#if NETFX_CORE
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.WebApiCompatShim/ContentNegotiator/FormattingUtilities.cs]: 🥇 to this guy for remaining hidden and / or ignored for so long ❗️

File: src/Microsoft.AspNetCore.Mvc.WebApiCompatShim/Microsoft.AspNetCore.Mvc.WebApiCompatShim.csproj [hunk 7]
```diff
@@ -6,11 +6,11 @@
     <Description>Provides compatibility in ASP.NET Core MVC with ASP.NET Web API 2 to simplify migration of existing Web API implementations.
 Commonly used types:
 System.Web.Http.ApiController</Description>
-    <TargetFrameworks>net451;netstandard1.6</TargetFrameworks>
+    <TargetFrameworks>net46;netstandard1.6</TargetFrameworks>
     <NoWarn>$(NoWarn);CS1591</NoWarn>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <PackageTags>aspnetcore;aspnetcoremvc;aspnetwebapi</PackageTags>
-    <PackageTargetFallback Condition=" '$(TargetFramework)' == 'netstandard1.6' ">$(PackageTargetFallback);portable-net451+win8</PackageTargetFallback>
+    <PackageTargetFallback Condition=" '$(TargetFramework)' == 'netstandard1.6' ">$(PackageTargetFallback);portable-net46+win8</PackageTargetFallback>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.WebApiCompatShim/Microsoft.AspNetCore.Mvc.WebApiCompatShim.csproj]: Doesn't changing the fallback TFM break anything? Either way, undo this change and similar changes a few places below.

File: test/Microsoft.AspNetCore.Mvc.Abstractions.Test/Microsoft.AspNetCore.Mvc.Abstractions.Test.csproj [hunk 8]
```diff
@@ -12,6 +12,7 @@
     <PackageReference Include="Microsoft.AspNetCore.Testing" Version="$(AspNetCoreVersion)" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="$(TestSdkVersion)" />
     <PackageReference Include="Moq" Version="$(MoqVersion)" />
+    <PackageReference Include="System.Reflection.TypeExtensions" Version="$(CoreFxVersion)" />
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Abstractions.Test/Microsoft.AspNetCore.Mvc.Abstractions.Test.csproj]: Why is this needed now? Same question for similar additions in (3 or so) other test projects. Should be brought in transitively in any case.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Microsoft.AspNetCore.Mvc.Razor.Host.Test.csproj [hunk 9]
```diff
@@ -3,12 +3,14 @@
...
-    <TargetFrameworks>netcoreapp1.1;net452</TargetFrameworks>
-    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netcoreapp1.1</TargetFrameworks>
+    <TargetFrameworks>netcoreapp2.0;net46</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netcoreapp2.0</TargetFrameworks>
     <PreserveCompilationContext>true</PreserveCompilationContext>
     <DefineConstants Condition="'$(GenerateBaselines)'=='true'">$(DefineConstants);GENERATE_BASELINES</DefineConstants>
     <DefineConstants>$(DefineConstants);__RemoveThisBitTo__GENERATE_BASELINES</DefineConstants>
     <DefaultItemExcludes>$(DefaultItemExcludes);TestFiles\**</DefaultItemExcludes>
+    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
+    <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Microsoft.AspNetCore.Mvc.Razor.Host.Test.csproj]: This project didn't previously need the binding redirects workaround. Shouldn't now. Revert this if possible. Same in 3 or so other projects with these additions.

File: src/Microsoft.AspNetCore.Mvc.WebApiCompatShim/HttpRequestMessage/HttpRequestMessageExtensions.cs [hunk 10]
```diff
@@ -19,7 +19,7 @@ namespace System.Net.Http
     /// </summary>
     public static class HttpRequestMessageExtensions
     {
-#if !NETSTANDARD1_6
+#if NET46
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.WebApiCompatShim/HttpRequestMessage/HttpRequestMessageExtensions.cs]: Like standardizing on the positive. 👍

File: test/Microsoft.AspNetCore.Mvc.WebApiCompatShimTest/DefaultContentNegotiatorTest.cs [hunk 11]
```diff
@@ -149,10 +149,13 @@ public class DefaultContentNegotiatorTests
...
             get
             {
-#if !NETCOREAPP1_1
+#if NET46
                 // Only mapping and accept makes sense with q != 1.0
                 MediaTypeFormatterMatch matchMapping10 = CreateMatch(1.0, MediaTypeFormatterMatchRanking.MatchOnRequestWithMediaTypeMapping);
                 MediaTypeFormatterMatch matchMapping05 = CreateMatch(0.5, MediaTypeFormatterMatchRanking.MatchOnRequestWithMediaTypeMapping);
+#elif NETCOREAPP2_0
+#else
+#error the target framework needs to be updated.
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.WebApiCompatShimTest/DefaultContentNegotiatorTest.cs]: What is the point of including this boilerplate multiple times in the same file?

File: test/WebSites/RazorWebSite/Services/FrameworkSpecificHelper.cs [hunk 12]
```diff
@@ -7,24 +7,26 @@ public class FrameworkSpecificHelper
...
+        public string ExecuteNET46Operation()
         {
-            return "This method is only defined in NET452";
+            return "This method is only defined in NET46";
         }
 #endif
 
-#if NETCOREAPP1_1_CUSTOM_DEFINE
+#if NETCOREAPP2_0_CUSTOM_DEFINE
         public string ExecuteNetCoreApp1_1Operation()
```
Comment (dougbu) [inline test/WebSites/RazorWebSite/Services/FrameworkSpecificHelper.cs]: Rename this method too.

File: src/Microsoft.AspNetCore.Mvc.WebApiCompatShim/Microsoft.AspNetCore.Mvc.WebApiCompatShim.csproj (hunk 7 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.WebApiCompatShim/Microsoft.AspNetCore.Mvc.WebApiCompatShim.csproj]: This was accidental. Bad find-replace fu

File: test/Microsoft.AspNetCore.Mvc.Abstractions.Test/Microsoft.AspNetCore.Mvc.Abstractions.Test.csproj (hunk 8 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Abstractions.Test/Microsoft.AspNetCore.Mvc.Abstractions.Test.csproj]: For some reason when I build Universe with everything net46, I didn't need this. But building Mvc when we have upstream dependencies cross-compiling for `net451` causes some weirdness. Other repos also had similar oddities. I've filed https://github.com/aspnet/Mvc/issues/6005 to remove this or at least figure out why we need this once I've updated everything in Universe

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Microsoft.AspNetCore.Mvc.Razor.Host.Test.csproj (hunk 9 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Microsoft.AspNetCore.Mvc.Razor.Host.Test.csproj]: This failed with a `MissingMemberException` due to Roslyn. Didn't you see similar issues with the functional test?

File: test/Microsoft.AspNetCore.Mvc.WebApiCompatShimTest/DefaultContentNegotiatorTest.cs (hunk 11 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.WebApiCompatShimTest/DefaultContentNegotiatorTest.cs]: The intent is for any `ifdef`s to follow this pattern so any of them fail if we miss updating them.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Microsoft.AspNetCore.Mvc.Razor.Host.Test.csproj (hunk 9 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Microsoft.AspNetCore.Mvc.Razor.Host.Test.csproj]: The functional tests had specific issues that led to needing binding redirects. Bizarre that Roslyn is now causing similar issues elsewhere. But, I guess we're using it in more tests now? If that's right, 🙈

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/EmptyArray.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/EmptyArray.cs]: 👍

File: src/Microsoft.AspNetCore.Mvc.ApiExplorer/Microsoft.AspNetCore.Mvc.ApiExplorer.csproj (hunk 3 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ApiExplorer/Microsoft.AspNetCore.Mvc.ApiExplorer.csproj]: We can't target anything lower than netstandard1.6 in most of our projects. We'll eventually replace `net46` -> `netstandard2.0`  in src projects.

File: samples/MvcSandbox/MvcSandbox.csproj (hunk 2 above)
Comment (pranavkm) [inline samples/MvcSandbox/MvcSandbox.csproj]: We don't do that in other repos. Let's do that separately if we want to also parameterize the TFMs

File: src/Microsoft.AspNetCore.Mvc.Formatters.Xml/Internal/FormattingUtilities.cs (hunk 5 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Xml/Internal/FormattingUtilities.cs]: With the exception of `#if` in usings, I think I got them all?

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Microsoft.AspNetCore.Mvc.Razor.Host.Test.csproj (hunk 9 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Microsoft.AspNetCore.Mvc.Razor.Host.Test.csproj]: Here's the exception stack trace without these flags:

```
Failed   Microsoft.AspNetCore.Mvc.Razor.RazorEngineTest.RazorEngine_InjectWithSemicolon_Runtime
Error Message:
 System.MissingMethodException : Method not found: 'System.Collections.Immutable.ImmutableArray`1<Byte> System.Reflection.Metadata.MetadataReader.GetBlobContent(System.Reflection.Metadata.BlobHandle)'.
Stack Trace:
   at Microsoft.CodeAnalysis.MetadataReaderExtensions.CreateAssemblyIdentityOrThrow(MetadataReader reader, Version version, AssemblyFlags flags, BlobHandle publicKey, StringHandle name, StringHandle culture, Boolean isReference)
   at Microsoft.CodeAnalysis.MetadataReaderExtensions.ReadAssemblyIdentityOrThrow(MetadataReader reader)
   at Microsoft.CodeAnalysis.PEAssembly..ctor(AssemblyMetadata owner, ImmutableArray`1 modules)

```

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.Formatters.Xml/Internal/FormattingUtilities.cs (hunk 5 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Xml/Internal/FormattingUtilities.cs]: See at least `NonDisposableStream`, `XmlDataContractSerializerOutputFormatter`, `CSharpCompiler` and `DefaultRoslynCompilationService`. I may have missed a few visible in this PR.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Microsoft.AspNetCore.Mvc.Razor.Host.Test.csproj (hunk 9 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Microsoft.AspNetCore.Mvc.Razor.Host.Test.csproj]: 🙈

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/Microsoft.AspNetCore.Mvc.FunctionalTests.csproj [hunk 13]
```diff
@@ -2,16 +2,16 @@
   <Import Project="..\..\build\common.props" />
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp1.1;net452</TargetFrameworks>
-    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netcoreapp1.1</TargetFrameworks>
-    <PackageTargetFallback Condition=" '$(TargetFramework)' == 'netcoreapp1.1' ">$(PackageTargetFallback);portable-net451+win8</PackageTargetFallback>
+    <TargetFrameworks>netcoreapp2.0;net46</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netcoreapp2.0</TargetFrameworks>
+    <PackageTargetFallback Condition=" '$(TargetFramework)' == 'netcoreapp2.0' ">$(PackageTargetFallback);portable-net46+win8</PackageTargetFallback>
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/Microsoft.AspNetCore.Mvc.FunctionalTests.csproj]: The `$(PackageTargetFallback)` value should not change here or the two other places it currently does.

File: src/Microsoft.AspNetCore.Mvc.Formatters.Xml/Internal/FormattingUtilities.cs (hunk 5 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Xml/Internal/FormattingUtilities.cs]: Sure, I'll fix these before I merge this in.

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/Microsoft.AspNetCore.Mvc.FunctionalTests.csproj (hunk 13 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/Microsoft.AspNetCore.Mvc.FunctionalTests.csproj]: 👍

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/EmptyArray.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/EmptyArray.cs]: This guy <---- is pretty stoked on getting to use `Array.Empty<>`

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/EmptyArray.cs]: `throw new AmbiguousRyanException{"this guy")`
